### PR TITLE
Linear Combination of Block Encodings

### PR DESCRIPTION
**Pull request by renezander90**
## Description

Reimplements block-encoding linear combinations as a single flattened LCU instead of
recursively nested binary combinations, and moves the arithmetic out of
`block_encoding_base.py` into a new `block_encoding_combination.py`.

Terms are held in a canonical, immutable tuple, so `A + B + C` builds one three-term
LCU rather than `(A + B) + C`. The mutually exclusive SELECT branches share a single
ancilla workspace sized to the largest term, including when the terms have different
ancilla layouts. Ancilla use therefore grows logarithmically in the term count instead
of linearly, at unchanged normalization.

## Related Issues

Related to #865

## Type of Change

- [x] Feature (new functionality)
- [x] Change Request (modification of existing functionality)
- [ ] Bug Fix
- [x] Refactoring (no behavior change)
- [x] Performance improvement
- [x] Documentation
- [ ] CI / Build

## Breaking Change?
- [ ] Yes
- [x] No

`+`, `-`, `*` and unary `-` keep their signatures and their encoded operator. `A + B`
now returns a `LinearCombinationBlockEncoding` rather than a plain `BlockEncoding`;
it is a subclass, so attribute and method access is unchanged.

## What was changed?

- Added `LinearCombinationBlockEncoding`, a `BlockEncoding` subclass holding a
  canonical immutable tuple of `(coefficient, block_encoding)` terms. Registered as a
  JAX pytree, and exported from `qrisp.block_encodings`.
- Reimplemented `__add__`, `__sub__`, `__mul__`, `__rmul__` and `__neg__` to build and
  merge these terms, so chained arithmetic flattens into one LCU instead of nesting.
- Added the `BlockEncoding.linear_combination` classmethod for building an LCU from an
  explicit list of block encodings and coefficients.
- Added `_AncillaLayout` (`ancilla_layout.py`), which records ancilla templates and
  their sizes and reconstructs typed views into a shared register. Sizes known at build
  time are kept as Python ints so the layout stays independent of any JAX trace.
- SELECT branches now take the shared workspace as an operand instead of capturing it,
  so they are built once ahead of tracing and reused on every invocation.
- Cached the derived state of a linear combination — unitary, SELECT branches, ancilla
  layouts, templates, coefficients and `alpha`. Jasp's caches are keyed on object
  identity, so re-deriving these handed every enclosing `q_switch` a fresh closure and
  re-traced the whole subtree once per nesting level.
- Kept build-time-known quantities as NumPy rather than `jnp` values. Inside a trace
  `jnp` returns a tracer even for constants, which silently downgraded `prepare` from a
  single opaque gate to a fully symbolic binary-tree construction.
- Moved all arithmetic out of `block_encoding_base.py` (−585 lines) into
  `block_encoding_combination.py`, attached to the class in `block_encoding.py`,
  matching the existing `constructors/` and `transformations/` assembler pattern.
- Fixed the import cycle this exposed: `transformations/pseudo_inv.py` now imports
  `BlockEncoding` from `block_encoding_base` rather than `block_encoding`. The
  `block_encodings` package is now cycle-free.
- Added a docs page for `linear_combination` and a changelog entry.

## How was it tested?

| Test-ID | Status |
|---------|--------|
| `tests/block_encodings_tests/` — full suite, 145 tests (118 on `main`) | ✅ |
| `test_block_encoding_arithmetic.py` — 25 tests (17 on `main`) | ✅ |
| `test_block_encoding_lcu_compilation.py` — 16 tests, new file | ✅ |
| `test_ancilla_layout.py` — 3 tests, new file | ✅ |
| `ruff check src/` on added lines + `ruff format --check` | ✅ |

Numerical equivalence is covered by comparing nested and flattened sums against
`BlockEncoding.from_operator` of the summed operator via `terminal_sampling`, across
several operator shapes and including terms with ancillas.

The new `test_block_encoding_lcu_compilation.py` asserts on compilation cost rather
than results, since a correct encoding can still be ruinously expensive to trace. It
pins down that nesting does not multiply tracing cost, that repeated application does
not re-trace, that `prepare` stays on its concrete path, and that cached state never
carries a tracer out of the trace that built it (`make_jaspr` → `count_ops` → `depth`
→ `num_qubits` on one reused object).

## Screenshots / Output (if applicable)

Peak qubits for a sum of *n* two-qubit operators, via `num_qubits`:

| terms | `main` | this PR | alpha |
|-------|--------|---------|-------|
| 2     | 3      | 4       | unchanged |
| 4     | 7      | 6       | unchanged |
| 8     | 15     | 8       | unchanged |

Linear growth becomes logarithmic. Two terms cost one qubit more, which is the
crossover of the shared-workspace construction.

## Checklist
- [x] My code follows the project's coding standards
- [x] I have performed a self-review
- [x] I have added/updated tests (referencing issue Test-IDs)
- [x] All tests pass locally and in CI
- [x] I have updated the documentation
- [x] I have added a changelog entry to `changelog-dev.rst`
- [x] Breaking changes are documented with migration path

## Reviewer Notes

- **The file split is the bulk of the diff.** `block_encoding_combination.py` is ~1000
  lines, but ~890 of them moved verbatim out of `block_encoding_base.py`. Reviewing the
  new logic means reading `LinearCombinationBlockEncoding`, `_AncillaLayout` and the
  caching; the arithmetic docstrings above them are unchanged apart from the receiver
  becoming an explicit `self`/`cls` parameter.
- **`# noqa: D417`** appears on six definitions. These are written as methods but
  defined at module level, and pydocstyle only exempts a receiver for functions inside a
  class body. The same false positive already exists unsuppressed on five functions in
  `transformations/` — worth deciding whether to handle those centrally in a follow-up.
- **Caching correctness is the sharp edge.** A cached value that captured a JAX tracer
  leaks it into the next transformation. Every cache is guarded on the value being
  trace-independent; `test_linear_combination_survives_reuse_across_transformations` is
  the test that catches getting this wrong.
- A follow-up PR adds the analogous `ProductBlockEncoding` for `@`.

### CODE REVIEW

File: src/qrisp/block_encodings/block_encoding_combination.py [hunk 1]
```diff
@@ -0,0 +1,1022 @@
...
+        terms.append((coefficient, block_encoding))
+
+    return cls._from_lcu_terms(terms)
+
+
+def build_from_lcu_terms(cls, terms: Sequence[_LCUTerm]) -> BlockEncoding:
+    """Build a linear-combination block encoding from weighted terms."""
+    terms = _canonicalize_lcu_terms(terms)
+    if len(terms) == 0:
+        raise ValueError("Cannot construct a block encoding from an all-zero linear combination.")
```
[the remark this answers is not in the document]
Comment (renezander90) [reply src/qrisp/block_encodings/block_encoding_combination.py]: Fixed in 85037dca8bbd86830f12bc6b3728aaaf807a6aaa

File: src/qrisp/block_encodings/block_encoding_combination.py [hunk 2]
```diff
@@ -0,0 +1,1008 @@
...
+            raise TypeError(f"Expected every item to be a BlockEncoding, but got {type(block_encoding).__name__}.")
+        terms.append((coefficient, block_encoding))
+
+    return cls._from_lcu_terms(terms)
+
+
+def build_from_lcu_terms(cls, terms: Sequence[_LCUTerm]) -> BlockEncoding:
+    """Build a linear-combination block encoding from weighted terms."""
+    num_ops = terms[0][1].num_ops if terms else 1
+    terms = _canonicalize_lcu_terms(terms)
```
[the remark this answers is not in the document]
Comment (renezander90) [reply src/qrisp/block_encodings/block_encoding_combination.py:124]: Fixed in bbe1093ab72951c430e99c5fe3f720ca26aafb45

File: src/qrisp/block_encodings/block_encoding_combination.py [hunk 3]
```diff
@@ -0,0 +1,1068 @@
...
+                # A coefficient that is real and positive needs no gate at all.
+                applies_phase = phase != 0.0
+
+            def unitary(*args):
+                child_unitary(*args)
+                if applies_phase:
+                    gphase(phase, args[0][0])
+
+            if not isinstance(coefficient, jax.core.Tracer) and _is_reusable_unitary(block_encoding):
+                object.__setattr__(self, "_cached_unitary", unitary)
```
[the remark this answers is not in the document]
Comment (renezander90) [reply src/qrisp/block_encodings/block_encoding_combination.py]: This is actually a bug in the `from_lcu` constructor and will be addressed in a separate PR. `alpha` should be nonnegative and real.
